Add book titles to search results

thavelick opened the pull request:
Implements/fixes #704 and fixes https://github.com/kiwix/kiwix-tools/issues/478

This is my first pull request here, so appologies if I've missed any process or coding style guidlines or best practices. Happy to revise anything or provide additional context.

It would be good as well to have a corresponding PR in kiwix/kiwix-tools and kiwix/kiwix-desktop to see how it works from a Kiwix port / end-user perspective.

thavelick commented:
I realized after the fact that by changing the constructor of `SearchRenderer`, this may break 3rd party software that uses libkiwix. I could see either:

1. Adding a new constructor that adds the libary but keeping the original one OR
2. Refactoring  this to set the library with a setter instead of with a constructor 

I happy to make one of the above changes but I want to make sure there's enough interest in this enhancement being merged in to the main rpeo before I spend a lot of time making it perfect. It works for my own purposes as is.

kelson42 commented:
@thavelick Thank you for your first PR. We will review it and give you a feedback soon.

veloman-yunkan commented on src/search_renderer.cpp:
```diff
@@ -90,6 +93,11 @@ std::string SearchRenderer::getHtml()
+    std::shared_ptr<zim::Archive> archive;
+    try {
+       archive = mp_library->getArchiveById(s.str());
+       result.set("bookTitle", archive ? getArchiveTitle(*archive) : "");
```
An empty string for `bookTitle` won't disable the mustache block marked with `{{#bookTitle}}...{{/bookTitle}}`. You will need to pass a `kainjow::mustache::data(false)` instead.

thavelick replied:
Updated that so it uses mustache's false.  Really this should be pertty much impossible. Both the `{#bookTitle}` wrapper and this null check are meant as guards against future bugs.

veloman-yunkan reviewed (changes requested):
The two commits should be combined into one.

Regarding the non-backward-compatible change in the public API of `SearchRenderer`, I personally wouldn't mind about it, but the final word should belong to @mgautierfr.

kelson42 commented:
@thavelick Any fix in view regarding @veloman-yunkan feedbacks?

thavelick commented:
@kelson42 I've been busy with work stuff and haven't had a chance to look the feedback in any depth yet. I think I should be able to address everything here this weekend.

thavelick commented:
@veloman-yunkan @kelson42 I've squashed my commits to one and addressed your comment

@mgautierfr looking forward to hearing your thoughts on the constructor.

thavelick commented:
Note, the ability to have mulitlpe seach results is currently a little hidden I've been testing this by starting up `kiwix-serve` and then hitting `/search?pattern=emulator` which will trigger resutls across archives if you have both wikipedia and wikihow zimfiles.

kelson42 commented:
@thavelick Thx for the update. It is true that the multizim search feature is not proeminent. This is because it does not scale in appropriate manner. This is a problem @maneeshpm is working on and I hope the problem will be fixed in one of the next minor releases.

thavelick commented on src/search_renderer.cpp:
```diff
@@ -90,6 +93,11 @@ std::string SearchRenderer::getHtml()
+    std::shared_ptr<zim::Archive> archive;
+    try {
+       archive = mp_library->getArchiveById(s.str());
+       result.set("bookTitle", archive ? getArchiveTitle(*archive) : kainjow::mustache::data(false));
```
@veloman-yunkan Your right, I don't think there is a situation in which archive can be null here. I've switched to just using `mp_library->getBookById(s.str()).getTitle()` as you've suggested

thavelick commented:
@veloman-yunkan @kelson42 Let me know if you need anything else form me to get this one over the line
